{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Steps:\n",
    "1. Make blocks from GFF\n",
    "2. Assign each SNP to a block\n",
    "3. Use indices of GT to make a DataFrame per block, and remove missing values\n",
    "4. Use sampling state to sample two haplotypes from each block DF\n",
    "5. Count number of segregating sites between blocks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "import allel\n",
    "import zarr\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (1) Make blocks from GFF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>block_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>73756</td>\n",
       "      <td>73857</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:73756-73857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>136255</td>\n",
       "      <td>136356</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:136255-136356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>237704</td>\n",
       "      <td>237805</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:237704-237805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>280192</td>\n",
       "      <td>280293</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:280192-280293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>352276</td>\n",
       "      <td>352377</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:352276-352377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483342</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_2</td>\n",
       "      <td>1661</td>\n",
       "      <td>1762</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_2:1661-1762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483447</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_5</td>\n",
       "      <td>11666</td>\n",
       "      <td>11767</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_5:11666-11767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483477</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_5</td>\n",
       "      <td>45471</td>\n",
       "      <td>45572</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_5:45471-45572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483621</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_8</td>\n",
       "      <td>40518</td>\n",
       "      <td>40619</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_8:40518-40619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483654</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_9</td>\n",
       "      <td>2180</td>\n",
       "      <td>2281</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_9:2180-2281</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4555 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        chr   start     end  \\\n",
       "35      brenthis_ino.SP_BI_364.chromosome_1   73756   73857   \n",
       "50      brenthis_ino.SP_BI_364.chromosome_1  136255  136356   \n",
       "127     brenthis_ino.SP_BI_364.chromosome_1  237704  237805   \n",
       "187     brenthis_ino.SP_BI_364.chromosome_1  280192  280293   \n",
       "359     brenthis_ino.SP_BI_364.chromosome_1  352276  352377   \n",
       "...                                     ...     ...     ...   \n",
       "483342    brenthis_ino.SP_BI_364.scaffold_2    1661    1762   \n",
       "483447    brenthis_ino.SP_BI_364.scaffold_5   11666   11767   \n",
       "483477    brenthis_ino.SP_BI_364.scaffold_5   45471   45572   \n",
       "483621    brenthis_ino.SP_BI_364.scaffold_8   40518   40619   \n",
       "483654    brenthis_ino.SP_BI_364.scaffold_9    2180    2281   \n",
       "\n",
       "                                                 block_id  \n",
       "35        brenthis_ino.SP_BI_364.chromosome_1:73756-73857  \n",
       "50      brenthis_ino.SP_BI_364.chromosome_1:136255-136356  \n",
       "127     brenthis_ino.SP_BI_364.chromosome_1:237704-237805  \n",
       "187     brenthis_ino.SP_BI_364.chromosome_1:280192-280293  \n",
       "359     brenthis_ino.SP_BI_364.chromosome_1:352276-352377  \n",
       "...                                                   ...  \n",
       "483342        brenthis_ino.SP_BI_364.scaffold_2:1661-1762  \n",
       "483447      brenthis_ino.SP_BI_364.scaffold_5:11666-11767  \n",
       "483477      brenthis_ino.SP_BI_364.scaffold_5:45471-45572  \n",
       "483621      brenthis_ino.SP_BI_364.scaffold_8:40518-40619  \n",
       "483654        brenthis_ino.SP_BI_364.scaffold_9:2180-2281  \n",
       "\n",
       "[4555 rows x 4 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff = pd.read_table(\"/Users/s2341012/Dropbox/DISMaL_chapter/brenthis_test/brenthis_data/brenthis_ino.SP_BI_364.v2_0.sequences.braker.gt.gff3\",\n",
    "               nrows=1000,\n",
    "               usecols=[0, 2, 3, 4],\n",
    "               sep=\"\\t\",\n",
    "               comment=\"#\",\n",
    "               header=None)\n",
    "\n",
    "gff.columns = [\"chr\", \"feature\", \"start\", \"end\"]\n",
    "\n",
    "def subset_introns(gff):\n",
    "    return gff[gff.iloc[:,1] == \"intron\"]\n",
    "\n",
    "def subset_length(gff, length):\n",
    "    return gff[gff.iloc[:,3] - gff.iloc[:,2] > length]\n",
    "\n",
    "def filter_ld(gff, ld_bp, sort=True):\n",
    "    #if sort:\n",
    "     #   gff = gff.sort_values(by=[\"start\", \"end\"])\n",
    "    \n",
    "    filtered = np.where([gff.iloc[row, -2] - gff.iloc[row-1, -1] > ld_bp for row in range(1, len(gff)-1)])\n",
    "\n",
    "    return gff.iloc[filtered[0], :]\n",
    "\n",
    "def trim_blocks(gff, length):\n",
    "    start_indices = [random.randint(gff.iloc[row, -2], (gff.iloc[row, -1]-length)) for row in range(0, len(gff))]\n",
    "    end_indices = np.array(start_indices) + length+1\n",
    "    return pd.DataFrame({\"chr\":gff.iloc[:,0], \"start\":start_indices, \"end\":end_indices})\n",
    "\n",
    "def make_block_indices(gff, blocklength, ld_bp, nrows=None):\n",
    "\n",
    "    gff = pd.read_table(gff,\n",
    "               nrows=nrows,\n",
    "               usecols=[0, 2, 3, 4],\n",
    "               sep=\"\\t\",\n",
    "               comment=\"#\",\n",
    "               header=None)\n",
    "\n",
    "    gff.columns = [\"chr\", \"feature\", \"start\", \"end\"]\n",
    "\n",
    "    introns = subset_introns(gff)\n",
    "    introns_filtered_len = subset_length(introns, blocklength)\n",
    "    introns_filtered_ld = filter_ld(introns_filtered_len, ld_bp)\n",
    "    block_indices = trim_blocks(introns_filtered_ld, blocklength)\n",
    "    block_indices[\"block_id\"] = block_indices[\"chr\"] + \":\" + block_indices[\"start\"].astype(str) + \"-\" + block_indices[\"end\"].astype(str)\n",
    "\n",
    "    return block_indices\n",
    "\n",
    "blocks = make_block_indices(\"/Users/s2341012/Dropbox/DISMaL_chapter/brenthis_test/brenthis_data/brenthis_ino.SP_BI_364.v2_0.sequences.braker.gt.gff3\", 100, 10000)\n",
    "blocks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (2) Assign each SNP to a block"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/s2341012/mambaforge/envs/dismal/lib/python3.11/site-packages/allel/io/vcf_read.py:1721: UserWarning: invalid FILTER header: '##FILTER=<ID=NonSnp,Description=\"Set if true: TYPE!=\\\\\"snp\\\\\"\">\\n'\n",
      "  warnings.warn('invalid FILTER header: %r' % header)\n"
     ]
    }
   ],
   "source": [
    "vcf_path = \"/Users/s2341012/Dropbox/DISMaL_chapter/brenthis_test/brenthis_data/brenthis_ino_daphne.vcf.gz\"\n",
    "zarr_path = \"/Users/s2341012/Dropbox/DISMaL_chapter/brenthis_test/brenthis_data/zarrstore\"\n",
    "\n",
    "try:\n",
    "    allel.vcf_to_zarr(vcf_path, zarr_path,\n",
    "                          fields=[\"samples\", \"calldata/GT\", \"variants/CHROM\", \"variants/POS\"], overwrite=False)\n",
    "except Exception: # if store already exists, proceed\n",
    "    pass \n",
    "\n",
    "callset = zarr.open_group(zarr_path, mode='r')\n",
    "chromosomes = callset[\"variants/CHROM\"][:]\n",
    "samples = callset[\"samples\"][:]\n",
    "gt = callset[\"calldata/GT\"][:]\n",
    "pos = callset[\"variants/POS\"][:]\n",
    "\n",
    "callset_positions = pd.DataFrame({\"chr\":chromosomes, \"pos\":pos})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>block_id</th>\n",
       "      <th>gt_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>73762</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:73756-73857</td>\n",
       "      <td>1596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>73762</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:73756-73857</td>\n",
       "      <td>1596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>73824</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:73756-73857</td>\n",
       "      <td>1597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>73824</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:73756-73857</td>\n",
       "      <td>1597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>73825</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:73756-73857</td>\n",
       "      <td>1598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40765</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_8</td>\n",
       "      <td>40617</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_8:40518-40619</td>\n",
       "      <td>22239363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40766</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_8</td>\n",
       "      <td>40617</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_8:40518-40619</td>\n",
       "      <td>22239363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40767</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_8</td>\n",
       "      <td>40584</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_8:40518-40619</td>\n",
       "      <td>22239359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40768</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_8</td>\n",
       "      <td>40586</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_8:40518-40619</td>\n",
       "      <td>22239360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40769</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_8</td>\n",
       "      <td>40618</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_8:40518-40619</td>\n",
       "      <td>22239364</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40770 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       chr    pos  \\\n",
       "0      brenthis_ino.SP_BI_364.chromosome_1  73762   \n",
       "1      brenthis_ino.SP_BI_364.chromosome_1  73762   \n",
       "2      brenthis_ino.SP_BI_364.chromosome_1  73824   \n",
       "3      brenthis_ino.SP_BI_364.chromosome_1  73824   \n",
       "4      brenthis_ino.SP_BI_364.chromosome_1  73825   \n",
       "...                                    ...    ...   \n",
       "40765    brenthis_ino.SP_BI_364.scaffold_8  40617   \n",
       "40766    brenthis_ino.SP_BI_364.scaffold_8  40617   \n",
       "40767    brenthis_ino.SP_BI_364.scaffold_8  40584   \n",
       "40768    brenthis_ino.SP_BI_364.scaffold_8  40586   \n",
       "40769    brenthis_ino.SP_BI_364.scaffold_8  40618   \n",
       "\n",
       "                                              block_id    gt_idx  \n",
       "0      brenthis_ino.SP_BI_364.chromosome_1:73756-73857      1596  \n",
       "1      brenthis_ino.SP_BI_364.chromosome_1:73756-73857      1596  \n",
       "2      brenthis_ino.SP_BI_364.chromosome_1:73756-73857      1597  \n",
       "3      brenthis_ino.SP_BI_364.chromosome_1:73756-73857      1597  \n",
       "4      brenthis_ino.SP_BI_364.chromosome_1:73756-73857      1598  \n",
       "...                                                ...       ...  \n",
       "40765    brenthis_ino.SP_BI_364.scaffold_8:40518-40619  22239363  \n",
       "40766    brenthis_ino.SP_BI_364.scaffold_8:40518-40619  22239363  \n",
       "40767    brenthis_ino.SP_BI_364.scaffold_8:40518-40619  22239359  \n",
       "40768    brenthis_ino.SP_BI_364.scaffold_8:40518-40619  22239360  \n",
       "40769    brenthis_ino.SP_BI_364.scaffold_8:40518-40619  22239364  \n",
       "\n",
       "[40770 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_snp_positions(chromosomes, pos, blocks):\n",
    "    \"\"\"Get DF [chr, pos, block, gt_idx]\"\"\"\n",
    "    positions = pd.DataFrame({\"chr\":chromosomes, \"pos\":pos})\n",
    "    positions[\"gt_idx\"] = positions.index\n",
    "\n",
    "    blocks_snps_df = pd.concat([positions[positions[\"pos\"].between(blocks.iloc[row, 1],\n",
    "                                                                blocks.iloc[row, 2])]\n",
    "                                                                  for row in range(len(blocks))])\n",
    "  \n",
    "    snp_pos_block = []\n",
    "\n",
    "    blocks_snps_df_ = blocks_snps_df.reset_index()\n",
    "\n",
    "    for row in range(len(blocks_snps_df_)):\n",
    "      position = blocks_snps_df_.iloc[row, 2]\n",
    "      snp_block = blocks[(blocks[\"start\"].astype(int) <= position) \n",
    "                         & (blocks[\"end\"].astype(int) >= position)][[\"chr\", \"block_id\"]]\n",
    "      snp_pos_block.append((snp_block.iloc[0][0], position, snp_block.iloc[0][1]))\n",
    "\n",
    "    snp_blocks = pd.DataFrame(snp_pos_block)\n",
    "    snp_blocks.columns = [\"chr\", \"pos\", \"block_id\"]\n",
    "\n",
    "    snp_blocks = snp_blocks.merge(positions, on=[\"chr\", \"pos\"])\n",
    "      \n",
    "    return snp_blocks\n",
    "\n",
    "get_snp_positions(chromosomes, pos, blocks)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (3) Use GT information to make a DF of calls per block"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>hap</th>\n",
       "      <th>1595</th>\n",
       "      <th>1596</th>\n",
       "      <th>1597</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SE_BI_1495</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SE_BI_1495</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FR_BI_1497</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FR_BI_1497</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FR_BD_1329</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FR_BD_1329</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>RS_BI_1496</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>RS_BI_1496</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RO_BD_956</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>RO_BD_956</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ES_BI_375</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ES_BI_375</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>UA_BI_1494</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>UA_BI_1494</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ES_BD_1141</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ES_BD_1141</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ES_BI_364</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ES_BI_364</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ES_BD_1489</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ES_BD_1489</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ES_BD_1490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ES_BD_1490</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>GR_BD_1491</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>GR_BD_1491</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>IT_BD_1493</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>IT_BD_1493</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample  hap  1595  1596  1597\n",
       "0   SE_BI_1495    0     0     0     0\n",
       "1   SE_BI_1495    1     0     0     0\n",
       "2   FR_BI_1497    0     0     0     0\n",
       "3   FR_BI_1497    1     0     0     1\n",
       "4   FR_BD_1329    0     0     1     1\n",
       "5   FR_BD_1329    1     0     1     1\n",
       "6   RS_BI_1496    0     0     0     0\n",
       "7   RS_BI_1496    1     1     0     1\n",
       "8    RO_BD_956    0     0     1     1\n",
       "9    RO_BD_956    1     0     1     1\n",
       "10   ES_BI_375    0     0     0     0\n",
       "11   ES_BI_375    1     0     0     0\n",
       "12  UA_BI_1494    0     0     0     0\n",
       "13  UA_BI_1494    1     0     0     1\n",
       "14  ES_BD_1141    0     0     1     1\n",
       "15  ES_BD_1141    1     0     1     1\n",
       "16   ES_BI_364    0     0     0     0\n",
       "17   ES_BI_364    1     0     0     0\n",
       "18  ES_BD_1489    0     0     1     1\n",
       "19  ES_BD_1489    1     0     1     1\n",
       "20  ES_BD_1490    0     0     1     1\n",
       "21  ES_BD_1490    1     0     1     1\n",
       "22  GR_BD_1491    0     0     1     1\n",
       "23  GR_BD_1491    1     0     1     1\n",
       "24  IT_BD_1493    0     0     1     1\n",
       "25  IT_BD_1493    1     0     1     1"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def block_callset_df(gt, gt_idxs, samples_list, ploidy):\n",
    "    block_calls = pd.DataFrame(list(itertools.product(samples_list, list(range(ploidy)))),\n",
    "                        columns=[\"sample\", \"hap\"])\n",
    "    \n",
    "    for idx in gt_idxs:\n",
    "        block_calls[idx] = gt[idx].flatten()\n",
    "\n",
    "    return block_calls.dropna()\n",
    "\n",
    "block_callset_df(gt, list(range(1595, 1598)), samples, 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def n_segr_sites(df):\n",
    "    # count segregating sites between samples of length 2\n",
    "    return sum(df.iloc[0, 2:] != df.iloc[1, 2:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def segregating_sites_spectrum(callset, snp_positions, ploidy, sampling_probabilities, samples_df, blocklen):\n",
    "\n",
    "    rng = np.random.default_rng()\n",
    "\n",
    "    populations = samples_df[\"population\"].unique()\n",
    "    pop1 = list(samples_df[\"sample\"][samples_df[\"population\"] == populations[0]])\n",
    "    pop2 = list(samples_df[\"sample\"][samples_df[\"population\"] == populations[1]])\n",
    "\n",
    "    sss = np.zeros(shape=(3, blocklen))\n",
    "\n",
    "    for block in snp_positions[\"block_id\"].unique():\n",
    "        block_gt_idxs = list(snp_positions[snp_positions[\"block_id\"] == block][\"gt_idx\"])\n",
    "        block_callset = block_callset_df(callset[\"calldata/GT\"],\n",
    "                                          block_gt_idxs,\n",
    "                                            callset[\"samples\"],\n",
    "                                              ploidy=ploidy)\n",
    "\n",
    "        sampling_state = int(np.where(rng.multinomial(n=1, pvals=sampling_probabilities) == 1)[0]) + 1\n",
    "\n",
    "        if sampling_state == 1:\n",
    "            block_sss = n_segr_sites(block_callset[block_callset[\"sample\"].isin(pop1)].sample(n=2))\n",
    "        elif sampling_state == 2:\n",
    "            block_sss = n_segr_sites(block_callset[block_callset[\"sample\"].isin(pop1)].sample(n=2))\n",
    "        else:\n",
    "            assert sampling_state == 3\n",
    "            block_sss = n_segr_sites(pd.concat([block_callset[block_callset[\"sample\"].isin(pop1)].sample(n=1),\n",
    "                                     block_callset[block_callset[\"sample\"].isin(pop1)].sample(n=2)]))\n",
    "\n",
    "        sss[sampling_state-1, int(block_sss)] = sss[sampling_state-1, int(block_sss)] + 1\n",
    "\n",
    "    return sss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SE_BI_1495',\n",
       " 'FR_BI_1497',\n",
       " 'FR_BD_1329',\n",
       " 'RS_BI_1496',\n",
       " 'RO_BD_956',\n",
       " 'ES_BI_375',\n",
       " 'UA_BI_1494',\n",
       " 'ES_BD_1141',\n",
       " 'ES_BI_364',\n",
       " 'ES_BD_1489',\n",
       " 'ES_BD_1490',\n",
       " 'GR_BD_1491',\n",
       " 'IT_BD_1493']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(callset[\"samples\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_df = pd.DataFrame(\n",
    "    {\n",
    "        \"sample\": callset[\"samples\"][:],\n",
    "        \"population\": [\"BI\", \"BI\", \"BD\", \"BI\", \"BD\", \"BI\", \"BI\", \"BD\", \"BI\", \"BD\", \"BD\", \"BD\", \"BD\"]\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "s3 = segregating_sites_spectrum(\n",
    "    callset,\n",
    "    snp_positions=get_snp_positions(callset[\"variants/CHROM\"][:], pos=callset[\"variants/POS\"][:], blocks=blocks),\n",
    "    ploidy=2,\n",
    "    sampling_probabilities=[0.25, 0.25, 0.5],\n",
    "    samples_df=samples_df,\n",
    "    blocklen=100\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[444., 263., 130., 104.,  57.,  24.,  23.,  11.,  12.,   2.,   3.,\n",
       "          1.,   0.,   1.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.],\n",
       "       [429., 266., 138.,  77.,  46.,  20.,  19.,  15.,   9.,   1.,   2.,\n",
       "          0.,   0.,   0.,   0.,   1.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.],\n",
       "       [896., 451., 269., 160.,  96.,  55.,  31.,  19.,   9.,   6.,   4.,\n",
       "          1.,   2.,   1.,   0.,   1.,   1.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.]])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s3"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dismal",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
