{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Steps:\n",
    "1. Make blocks from GFF\n",
    "2. Assign each SNP to a block\n",
    "3. Use indices of GT to make a DataFrame per block, and remove missing values\n",
    "4. Use sampling state to sample two haplotypes from each block DF\n",
    "5. Count number of segregating sites between blocks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "import allel\n",
    "import zarr\n",
    "import itertools\n",
    "import requests\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (1) Make blocks from GFF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>gene</td>\n",
       "      <td>18787</td>\n",
       "      <td>19695</td>\n",
       "      <td>0.60</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>18787</td>\n",
       "      <td>19695</td>\n",
       "      <td>0.60</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>18787</td>\n",
       "      <td>18789</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>18790</td>\n",
       "      <td>19695</td>\n",
       "      <td>0.60</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>exon</td>\n",
       "      <td>18790</td>\n",
       "      <td>19695</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483674</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_9</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>exon</td>\n",
       "      <td>46784</td>\n",
       "      <td>47186</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483675</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_9</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>intron</td>\n",
       "      <td>47187</td>\n",
       "      <td>53566</td>\n",
       "      <td>0.88</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483676</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_9</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>53567</td>\n",
       "      <td>54497</td>\n",
       "      <td>0.89</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483677</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_9</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>exon</td>\n",
       "      <td>53567</td>\n",
       "      <td>54497</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483678</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_9</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>54495</td>\n",
       "      <td>54497</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>483679 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      seqid    source         type  start  \\\n",
       "0       brenthis_ino.SP_BI_364.chromosome_1  AUGUSTUS         gene  18787   \n",
       "1       brenthis_ino.SP_BI_364.chromosome_1  AUGUSTUS         mRNA  18787   \n",
       "2       brenthis_ino.SP_BI_364.chromosome_1  AUGUSTUS   stop_codon  18787   \n",
       "3       brenthis_ino.SP_BI_364.chromosome_1  AUGUSTUS          CDS  18790   \n",
       "4       brenthis_ino.SP_BI_364.chromosome_1  AUGUSTUS         exon  18790   \n",
       "...                                     ...       ...          ...    ...   \n",
       "483674    brenthis_ino.SP_BI_364.scaffold_9  AUGUSTUS         exon  46784   \n",
       "483675    brenthis_ino.SP_BI_364.scaffold_9  AUGUSTUS       intron  47187   \n",
       "483676    brenthis_ino.SP_BI_364.scaffold_9  AUGUSTUS          CDS  53567   \n",
       "483677    brenthis_ino.SP_BI_364.scaffold_9  AUGUSTUS         exon  53567   \n",
       "483678    brenthis_ino.SP_BI_364.scaffold_9  AUGUSTUS  start_codon  54495   \n",
       "\n",
       "          end  score strand  phase  \n",
       "0       19695   0.60      -     -1  \n",
       "1       19695   0.60      -     -1  \n",
       "2       18789  -1.00      -      0  \n",
       "3       19695   0.60      -      0  \n",
       "4       19695  -1.00      -     -1  \n",
       "...       ...    ...    ...    ...  \n",
       "483674  47186  -1.00      -     -1  \n",
       "483675  53566   0.88      -     -1  \n",
       "483676  54497   0.89      -      0  \n",
       "483677  54497  -1.00      -     -1  \n",
       "483678  54497  -1.00      -      0  \n",
       "\n",
       "[483679 rows x 8 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff3 = allel.gff3_to_dataframe(\"/Users/s2341012/Dropbox/DISMaL_chapter/brenthis_test/brenthis_data/brenthis_ino.SP_BI_364.v2_0.sequences.braker.gt.gff3\")\n",
    "gff3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_ld(df, ld_bp):\n",
    "    filtered = np.where([df.iloc[row, 1] - df.iloc[row-1, 2] > ld_bp for row in range(1, len(df)-1)])\n",
    "    return df.iloc[filtered[0], :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>67375</td>\n",
       "      <td>67476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>68439</td>\n",
       "      <td>68540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>69411</td>\n",
       "      <td>69512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>70046</td>\n",
       "      <td>70147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>71017</td>\n",
       "      <td>71118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>920</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>599278</td>\n",
       "      <td>599379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>929</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>601570</td>\n",
       "      <td>601671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>932</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>602531</td>\n",
       "      <td>602632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>935</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>602921</td>\n",
       "      <td>603022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>938</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>603649</td>\n",
       "      <td>603750</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>257 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     chr   start     end\n",
       "17   brenthis_ino.SP_BI_364.chromosome_1   67375   67476\n",
       "20   brenthis_ino.SP_BI_364.chromosome_1   68439   68540\n",
       "23   brenthis_ino.SP_BI_364.chromosome_1   69411   69512\n",
       "26   brenthis_ino.SP_BI_364.chromosome_1   70046   70147\n",
       "29   brenthis_ino.SP_BI_364.chromosome_1   71017   71118\n",
       "..                                   ...     ...     ...\n",
       "920  brenthis_ino.SP_BI_364.chromosome_1  599278  599379\n",
       "929  brenthis_ino.SP_BI_364.chromosome_1  601570  601671\n",
       "932  brenthis_ino.SP_BI_364.chromosome_1  602531  602632\n",
       "935  brenthis_ino.SP_BI_364.chromosome_1  602921  603022\n",
       "938  brenthis_ino.SP_BI_364.chromosome_1  603649  603750\n",
       "\n",
       "[257 rows x 3 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff1000 = allel.gff3_to_dataframe(\"../test_data/brenthis_head1000.gff3\")\n",
    "introns_df = gff1000[gff1000[\"type\"] == \"intron\"]\n",
    "blocklen=100\n",
    "trim_blocks(introns_df[(introns_df[\"end\"] - introns_df[\"start\"]) > blocklen], blocklen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _trim_blocks(df, blocklen):\n",
    "    start_indices = [random.randint(df.iloc[row, 3], (df.iloc[row, 4]-blocklen)) for row in range(0, len(df))]\n",
    "    end_indices = np.array(start_indices) + blocklen+1\n",
    "    return pd.DataFrame({\"chr\":df.iloc[:,0], \"start\":start_indices, \"end\":end_indices})\n",
    "\n",
    "def _filter_ld(df, ld_bp):\n",
    "    sorted = df.sort_values([\"chr\", \"start\"])\n",
    "\n",
    "    filtered_dfs = []\n",
    "    for chromosome in df[\"chr\"].unique():\n",
    "        df_chr = sorted[sorted[\"chr\"] == chromosome]\n",
    "        filtered = np.where([sorted.iloc[row, 1] - df_chr.iloc[row-1, 2] > ld_bp for row in range(1, len(df_chr)-1)])\n",
    "        filtered_dfs.append(df_chr.iloc[filtered[0], :])\n",
    "    \n",
    "    return pd.concat(filtered_dfs)\n",
    "\n",
    "\n",
    "def blocks_from_gff3(gff3_path, blocklen, ld_dist_bp):\n",
    "    \"\"\"Subset GFF3 file for intronic blocks of length blocklen, at least ld_dist apart\"\"\"\n",
    "\n",
    "    pd.set_option('chained_assignment',None)\n",
    "    \n",
    "    gff = allel.gff3_to_dataframe(gff3_path)\n",
    "    \n",
    "    # Subset intronic regions\n",
    "    introns_df = gff[gff[\"type\"] == \"intron\"]\n",
    "\n",
    "    # Subset len > blocklen, and trim to blocklen\n",
    "    blocks_df = _trim_blocks(introns_df[(introns_df[\"end\"] - introns_df[\"start\"]) > blocklen], blocklen)\n",
    "\n",
    "    # Filter for LD\n",
    "    blocks_ld_filtered = _filter_ld(blocks_df, ld_dist_bp)\n",
    "\n",
    "    # Include block information in output df\n",
    "    blocks_ld_filtered.loc[:, \"block_id\"] = blocks_ld_filtered.loc[:,\"chr\"] + \":\" + blocks_ld_filtered.loc[:,\"start\"].astype(str) + \"-\" + blocks_ld_filtered.loc[:,\"end\"].astype(str)\n",
    "\n",
    "    return blocks_ld_filtered\n",
    "\n",
    "\n",
    "gff_df = blocks_from_gff3(\"../test_data/brenthis_head1000.gff3\",\n",
    "            blocklen=100, ld_dist_bp=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>block_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>73731</td>\n",
       "      <td>73832</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:73731-73832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>126210</td>\n",
       "      <td>126311</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:126210-126311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>141537</td>\n",
       "      <td>141638</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:141537-141638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>237755</td>\n",
       "      <td>237856</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:237755-237856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>280458</td>\n",
       "      <td>280559</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:280458-280559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>337149</td>\n",
       "      <td>337250</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:337149-337250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>351434</td>\n",
       "      <td>351535</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:351434-351535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>365089</td>\n",
       "      <td>365190</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:365089-365190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>419328</td>\n",
       "      <td>419429</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:419328-419429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>589</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>469663</td>\n",
       "      <td>469764</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:469663-469764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>687</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>481371</td>\n",
       "      <td>481472</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:481371-481472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>696</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>530020</td>\n",
       "      <td>530121</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:530020-530121</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     chr   start     end  \\\n",
       "35   brenthis_ino.SP_BI_364.chromosome_1   73731   73832   \n",
       "47   brenthis_ino.SP_BI_364.chromosome_1  126210  126311   \n",
       "50   brenthis_ino.SP_BI_364.chromosome_1  141537  141638   \n",
       "127  brenthis_ino.SP_BI_364.chromosome_1  237755  237856   \n",
       "187  brenthis_ino.SP_BI_364.chromosome_1  280458  280559   \n",
       "353  brenthis_ino.SP_BI_364.chromosome_1  337149  337250   \n",
       "359  brenthis_ino.SP_BI_364.chromosome_1  351434  351535   \n",
       "368  brenthis_ino.SP_BI_364.chromosome_1  365089  365190   \n",
       "508  brenthis_ino.SP_BI_364.chromosome_1  419328  419429   \n",
       "589  brenthis_ino.SP_BI_364.chromosome_1  469663  469764   \n",
       "687  brenthis_ino.SP_BI_364.chromosome_1  481371  481472   \n",
       "696  brenthis_ino.SP_BI_364.chromosome_1  530020  530121   \n",
       "\n",
       "                                              block_id  \n",
       "35     brenthis_ino.SP_BI_364.chromosome_1:73731-73832  \n",
       "47   brenthis_ino.SP_BI_364.chromosome_1:126210-126311  \n",
       "50   brenthis_ino.SP_BI_364.chromosome_1:141537-141638  \n",
       "127  brenthis_ino.SP_BI_364.chromosome_1:237755-237856  \n",
       "187  brenthis_ino.SP_BI_364.chromosome_1:280458-280559  \n",
       "353  brenthis_ino.SP_BI_364.chromosome_1:337149-337250  \n",
       "359  brenthis_ino.SP_BI_364.chromosome_1:351434-351535  \n",
       "368  brenthis_ino.SP_BI_364.chromosome_1:365089-365190  \n",
       "508  brenthis_ino.SP_BI_364.chromosome_1:419328-419429  \n",
       "589  brenthis_ino.SP_BI_364.chromosome_1:469663-469764  \n",
       "687  brenthis_ino.SP_BI_364.chromosome_1:481371-481472  \n",
       "696  brenthis_ino.SP_BI_364.chromosome_1:530020-530121  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[58875, 105057, 42817, 56447, 15058, 12910, 58498, 45001, 12193, 47381]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "block_dist = [gff_df.iloc[row, 1]-gff_df.iloc[row-1, 2] for row in range(1, len(gff_df))]\n",
    "block_dist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[58875, 105057, 42817, 56447, 15058, 12910, 58498, 45001, 12193, 47381]\n"
     ]
    }
   ],
   "source": [
    "distance_between_blocks = [gff_df.iloc[row, 1] -\n",
    "                               gff_df.iloc[row-1, 2] for row in range(1, len(gff_df))]\n",
    "\n",
    "assert (np.array(distance_between_blocks) > 10000).all()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'Response' object has no attribute 'endswith'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[18], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m subset_gff3(requests\u001b[39m.\u001b[39mget(\u001b[39m\"\u001b[39m\u001b[39mhttps://www.dropbox.com/scl/fi/wyew115q4j4xqh5kh973r/brenthis_ino.SP_BI_364.v2_0.sequences.braker.gt.gff3?rlkey=sq4tmjgmtuhx54v6iql9o8e8t&dl=0\u001b[39m\u001b[39m\"\u001b[39m),\n\u001b[1;32m      2\u001b[0m             \u001b[39m100\u001b[39m, \u001b[39m1000\u001b[39m)\n",
      "Cell \u001b[0;32mIn[17], line 10\u001b[0m, in \u001b[0;36msubset_gff3\u001b[0;34m(gff3_path, blocklen, ld_dist)\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39msubset_gff3\u001b[39m(gff3_path, blocklen, ld_dist):\n\u001b[1;32m      8\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"Subset GFF3 file for intronic blocks of length blocklen, at least ld_dist apart\"\"\"\u001b[39;00m\n\u001b[0;32m---> 10\u001b[0m     gff \u001b[39m=\u001b[39m allel\u001b[39m.\u001b[39mgff3_to_dataframe(gff3_path)\n\u001b[1;32m     12\u001b[0m     \u001b[39m# Subset intronic regions\u001b[39;00m\n\u001b[1;32m     13\u001b[0m     introns_df \u001b[39m=\u001b[39m gff[gff[\u001b[39m\"\u001b[39m\u001b[39mtype\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mintron\u001b[39m\u001b[39m\"\u001b[39m]\n",
      "File \u001b[0;32m~/mambaforge/envs/dismal/lib/python3.11/site-packages/allel/io/gff.py:209\u001b[0m, in \u001b[0;36mgff3_to_dataframe\u001b[0;34m(path, attributes, region, score_fill, phase_fill, attributes_fill, tabix, **kwargs)\u001b[0m\n\u001b[1;32m    206\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mpandas\u001b[39;00m\n\u001b[1;32m    208\u001b[0m \u001b[39m# read records\u001b[39;00m\n\u001b[0;32m--> 209\u001b[0m recs \u001b[39m=\u001b[39m \u001b[39mlist\u001b[39m(iter_gff3(path, attributes\u001b[39m=\u001b[39mattributes, region\u001b[39m=\u001b[39mregion,\n\u001b[1;32m    210\u001b[0m                       score_fill\u001b[39m=\u001b[39mscore_fill, phase_fill\u001b[39m=\u001b[39mphase_fill,\n\u001b[1;32m    211\u001b[0m                       attributes_fill\u001b[39m=\u001b[39mattributes_fill, tabix\u001b[39m=\u001b[39mtabix))\n\u001b[1;32m    213\u001b[0m \u001b[39m# load into pandas\u001b[39;00m\n\u001b[1;32m    214\u001b[0m columns \u001b[39m=\u001b[39m [\u001b[39m'\u001b[39m\u001b[39mseqid\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39msource\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mtype\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mstart\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mend\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mscore\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mstrand\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mphase\u001b[39m\u001b[39m'\u001b[39m]\n",
      "File \u001b[0;32m~/mambaforge/envs/dismal/lib/python3.11/site-packages/allel/io/gff.py:70\u001b[0m, in \u001b[0;36miter_gff3\u001b[0;34m(path, attributes, region, score_fill, phase_fill, attributes_fill, tabix)\u001b[0m\n\u001b[1;32m     68\u001b[0m     cmd \u001b[39m=\u001b[39m [tabix, path, region]\n\u001b[1;32m     69\u001b[0m     buffer \u001b[39m=\u001b[39m subprocess\u001b[39m.\u001b[39mPopen(cmd, stdout\u001b[39m=\u001b[39msubprocess\u001b[39m.\u001b[39mPIPE)\u001b[39m.\u001b[39mstdout\n\u001b[0;32m---> 70\u001b[0m \u001b[39melif\u001b[39;00m path\u001b[39m.\u001b[39mendswith(\u001b[39m'\u001b[39m\u001b[39m.gz\u001b[39m\u001b[39m'\u001b[39m) \u001b[39mor\u001b[39;00m path\u001b[39m.\u001b[39mendswith(\u001b[39m'\u001b[39m\u001b[39m.bgz\u001b[39m\u001b[39m'\u001b[39m):\n\u001b[1;32m     71\u001b[0m     buffer \u001b[39m=\u001b[39m gzip\u001b[39m.\u001b[39mopen(path, mode\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39mrb\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[1;32m     72\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'Response' object has no attribute 'endswith'"
     ]
    }
   ],
   "source": [
    "subset_gff3(requests.get(\"https://www.dropbox.com/scl/fi/wyew115q4j4xqh5kh973r/brenthis_ino.SP_BI_364.v2_0.sequences.braker.gt.gff3?rlkey=sq4tmjgmtuhx54v6iql9o8e8t&dl=0\"),\n",
    "            100, 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>block_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>73671</td>\n",
       "      <td>73772</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:73671-73772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>143980</td>\n",
       "      <td>144081</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:143980-144081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>237688</td>\n",
       "      <td>237789</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:237688-237789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>280200</td>\n",
       "      <td>280301</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:280200-280301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>352222</td>\n",
       "      <td>352323</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:352222-352323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483342</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_2</td>\n",
       "      <td>1972</td>\n",
       "      <td>2073</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_2:1972-2073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483447</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_5</td>\n",
       "      <td>12075</td>\n",
       "      <td>12176</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_5:12075-12176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483477</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_5</td>\n",
       "      <td>45474</td>\n",
       "      <td>45575</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_5:45474-45575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483621</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_8</td>\n",
       "      <td>40533</td>\n",
       "      <td>40634</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_8:40533-40634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483654</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_9</td>\n",
       "      <td>2296</td>\n",
       "      <td>2397</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_9:2296-2397</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4555 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        chr   start     end  \\\n",
       "35      brenthis_ino.SP_BI_364.chromosome_1   73671   73772   \n",
       "50      brenthis_ino.SP_BI_364.chromosome_1  143980  144081   \n",
       "127     brenthis_ino.SP_BI_364.chromosome_1  237688  237789   \n",
       "187     brenthis_ino.SP_BI_364.chromosome_1  280200  280301   \n",
       "359     brenthis_ino.SP_BI_364.chromosome_1  352222  352323   \n",
       "...                                     ...     ...     ...   \n",
       "483342    brenthis_ino.SP_BI_364.scaffold_2    1972    2073   \n",
       "483447    brenthis_ino.SP_BI_364.scaffold_5   12075   12176   \n",
       "483477    brenthis_ino.SP_BI_364.scaffold_5   45474   45575   \n",
       "483621    brenthis_ino.SP_BI_364.scaffold_8   40533   40634   \n",
       "483654    brenthis_ino.SP_BI_364.scaffold_9    2296    2397   \n",
       "\n",
       "                                                 block_id  \n",
       "35        brenthis_ino.SP_BI_364.chromosome_1:73671-73772  \n",
       "50      brenthis_ino.SP_BI_364.chromosome_1:143980-144081  \n",
       "127     brenthis_ino.SP_BI_364.chromosome_1:237688-237789  \n",
       "187     brenthis_ino.SP_BI_364.chromosome_1:280200-280301  \n",
       "359     brenthis_ino.SP_BI_364.chromosome_1:352222-352323  \n",
       "...                                                   ...  \n",
       "483342        brenthis_ino.SP_BI_364.scaffold_2:1972-2073  \n",
       "483447      brenthis_ino.SP_BI_364.scaffold_5:12075-12176  \n",
       "483477      brenthis_ino.SP_BI_364.scaffold_5:45474-45575  \n",
       "483621      brenthis_ino.SP_BI_364.scaffold_8:40533-40634  \n",
       "483654        brenthis_ino.SP_BI_364.scaffold_9:2296-2397  \n",
       "\n",
       "[4555 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff = pd.read_table(\"/Users/s2341012/Dropbox/DISMaL_chapter/brenthis_test/brenthis_data/brenthis_ino.SP_BI_364.v2_0.sequences.braker.gt.gff3\",\n",
    "               nrows=1000,\n",
    "               usecols=[0, 2, 3, 4],\n",
    "               sep=\"\\t\",\n",
    "               comment=\"#\",\n",
    "               header=None)\n",
    "\n",
    "gff.columns = [\"chr\", \"feature\", \"start\", \"end\"]\n",
    "\n",
    "def subset_introns(gff):\n",
    "    return gff[gff.iloc[:,1] == \"intron\"]\n",
    "\n",
    "def subset_length(gff, length):\n",
    "    return gff[gff.iloc[:,3] - gff.iloc[:,2] > length]\n",
    "\n",
    "def filter_ld(gff, ld_bp, sort=True):\n",
    "    #if sort:\n",
    "     #   gff = gff.sort_values(by=[\"start\", \"end\"])\n",
    "    \n",
    "    filtered = np.where([gff.iloc[row, -2] - gff.iloc[row-1, -1] > ld_bp for row in range(1, len(gff)-1)])\n",
    "\n",
    "    return gff.iloc[filtered[0], :]\n",
    "\n",
    "def trim_blocks(gff, length):\n",
    "    start_indices = [random.randint(gff.iloc[row, -2], (gff.iloc[row, -1]-length)) for row in range(0, len(gff))]\n",
    "    end_indices = np.array(start_indices) + length+1\n",
    "    return pd.DataFrame({\"chr\":gff.iloc[:,0], \"start\":start_indices, \"end\":end_indices})\n",
    "\n",
    "def make_block_indices(gff, blocklength, ld_bp, nrows=None):\n",
    "\n",
    "    gff = pd.read_table(gff,\n",
    "               nrows=nrows,\n",
    "               usecols=[0, 2, 3, 4],\n",
    "               sep=\"\\t\",\n",
    "               comment=\"#\",\n",
    "               header=None)\n",
    "\n",
    "    gff.columns = [\"chr\", \"feature\", \"start\", \"end\"]\n",
    "\n",
    "    introns = subset_introns(gff)\n",
    "    introns_filtered_len = subset_length(introns, blocklength)\n",
    "    introns_filtered_ld = filter_ld(introns_filtered_len, ld_bp)\n",
    "    block_indices = trim_blocks(introns_filtered_ld, blocklength)\n",
    "    block_indices[\"block_id\"] = block_indices[\"chr\"] + \":\" + block_indices[\"start\"].astype(str) + \"-\" + block_indices[\"end\"].astype(str)\n",
    "\n",
    "    return block_indices\n",
    "\n",
    "blocks = make_block_indices(\"/Users/s2341012/Dropbox/DISMaL_chapter/brenthis_test/brenthis_data/brenthis_ino.SP_BI_364.v2_0.sequences.braker.gt.gff3\", 100, 10000)\n",
    "blocks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (2) Assign each SNP to a block"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/s2341012/mambaforge/envs/dismal/lib/python3.11/site-packages/allel/io/vcf_read.py:1721: UserWarning: invalid FILTER header: '##FILTER=<ID=NonSnp,Description=\"Set if true: TYPE!=\\\\\"snp\\\\\"\">\\n'\n",
      "  warnings.warn('invalid FILTER header: %r' % header)\n"
     ]
    }
   ],
   "source": [
    "vcf_path = \"/Users/s2341012/Dropbox/DISMaL_chapter/brenthis_test/brenthis_data/brenthis_ino_daphne.vcf.gz\"\n",
    "zarr_path = \"/Users/s2341012/Dropbox/DISMaL_chapter/brenthis_test/brenthis_data/zarrstore\"\n",
    "\n",
    "try:\n",
    "    allel.vcf_to_zarr(vcf_path, zarr_path,\n",
    "                          fields=[\"samples\", \"calldata/GT\", \"variants/CHROM\", \"variants/POS\"], overwrite=False)\n",
    "except Exception: # if store already exists, proceed\n",
    "    pass \n",
    "\n",
    "callset = zarr.open_group(zarr_path, mode='r')\n",
    "chromosomes = callset[\"variants/CHROM\"][:]\n",
    "samples = callset[\"samples\"][:]\n",
    "gt = callset[\"calldata/GT\"][:]\n",
    "pos = callset[\"variants/POS\"][:]\n",
    "\n",
    "callset_positions = pd.DataFrame({\"chr\":chromosomes, \"pos\":pos})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "class CallSet:\n",
    "    \"\"\"Class to represent a skallele callset (i.e. a VCF)\"\"\"\n",
    "\n",
    "    def __init__(self, vcf_path=None, zarr_path=None):\n",
    "        \n",
    "        if zarr_path is None:\n",
    "            self.zarr_path = \"zarrstore\"\n",
    "        else:\n",
    "            self.zarr_path = zarr_path\n",
    "\n",
    "        \n",
    "        if vcf_path is None:\n",
    "            assert zarr_path is not None\n",
    "\n",
    "        self.vcf_path = vcf_path\n",
    "        if not os.path.exists(self.zarr_path):\n",
    "            allel.vcf_to_zarr(self.vcf_path, self.zarr_path,\n",
    "                          fields=[\"samples\", \"calldata/GT\", \"variants/CHROM\", \"variants/POS\"], overwrite=False)\n",
    "            \n",
    "        \n",
    "        self.callset = zarr.open_group(zarr_path, mode='r')\n",
    "        self.chromosomes = self.callset[\"variants/CHROM\"][:]\n",
    "        self.samples = self.callset[\"samples\"][:]\n",
    "        self.gt = self.callset[\"calldata/GT\"][:]\n",
    "        self.pos = self.callset[\"variants/POS\"][:]\n",
    "        self.callset_positions_df = pd.DataFrame({\"chr\":self.chromosomes, \"pos\":self.pos})\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "callset = CallSet(vcf_path=\"../test_data/brenthis_500K.vcf.gz\", \n",
    "        zarr_path=\"../test_data/zarrstore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zarr_path = \"/Users/s2341012/Dropbox/DISMaL_chapter/brenthis_test/brenthis_data/zarrstore\"\n",
    "os.path.exists(zarr_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>block_id</th>\n",
       "      <th>gt_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>73671</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:73671-73772</td>\n",
       "      <td>1594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>73696</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:73671-73772</td>\n",
       "      <td>1595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>73762</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:73671-73772</td>\n",
       "      <td>1596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>143980</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:143980-144081</td>\n",
       "      <td>6736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>144001</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:143980-144081</td>\n",
       "      <td>6737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>469439</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:469429-469530</td>\n",
       "      <td>31056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>469461</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:469429-469530</td>\n",
       "      <td>31057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>469481</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:469429-469530</td>\n",
       "      <td>31058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>469487</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:469429-469530</td>\n",
       "      <td>31059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>469525</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:469429-469530</td>\n",
       "      <td>31060</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    chr     pos  \\\n",
       "0   brenthis_ino.SP_BI_364.chromosome_1   73671   \n",
       "1   brenthis_ino.SP_BI_364.chromosome_1   73696   \n",
       "2   brenthis_ino.SP_BI_364.chromosome_1   73762   \n",
       "3   brenthis_ino.SP_BI_364.chromosome_1  143980   \n",
       "4   brenthis_ino.SP_BI_364.chromosome_1  144001   \n",
       "..                                  ...     ...   \n",
       "57  brenthis_ino.SP_BI_364.chromosome_1  469439   \n",
       "58  brenthis_ino.SP_BI_364.chromosome_1  469461   \n",
       "59  brenthis_ino.SP_BI_364.chromosome_1  469481   \n",
       "60  brenthis_ino.SP_BI_364.chromosome_1  469487   \n",
       "61  brenthis_ino.SP_BI_364.chromosome_1  469525   \n",
       "\n",
       "                                             block_id  gt_idx  \n",
       "0     brenthis_ino.SP_BI_364.chromosome_1:73671-73772    1594  \n",
       "1     brenthis_ino.SP_BI_364.chromosome_1:73671-73772    1595  \n",
       "2     brenthis_ino.SP_BI_364.chromosome_1:73671-73772    1596  \n",
       "3   brenthis_ino.SP_BI_364.chromosome_1:143980-144081    6736  \n",
       "4   brenthis_ino.SP_BI_364.chromosome_1:143980-144081    6737  \n",
       "..                                                ...     ...  \n",
       "57  brenthis_ino.SP_BI_364.chromosome_1:469429-469530   31056  \n",
       "58  brenthis_ino.SP_BI_364.chromosome_1:469429-469530   31057  \n",
       "59  brenthis_ino.SP_BI_364.chromosome_1:469429-469530   31058  \n",
       "60  brenthis_ino.SP_BI_364.chromosome_1:469429-469530   31059  \n",
       "61  brenthis_ino.SP_BI_364.chromosome_1:469429-469530   31060  \n",
       "\n",
       "[62 rows x 4 columns]"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_block_snps(callset, blocks):\n",
    "    \"\"\"Create DataFrame of SNPs and which block they belong to [chr, pos, block, gt_idx]\"\"\"\n",
    "    positions = callset.callset_positions_df\n",
    "    positions[\"gt_idx\"] = positions.index\n",
    "\n",
    "    blocks_snps_df = pd.concat([positions[positions[\"pos\"].between(blocks.iloc[row, 1],\n",
    "                                                                blocks.iloc[row, 2])]\n",
    "                                                                  for row in range(len(blocks))])\n",
    "  \n",
    "    snp_pos_block = []\n",
    "\n",
    "    blocks_snps_df_ = blocks_snps_df.reset_index()\n",
    "\n",
    "    for row in range(len(blocks_snps_df_)):\n",
    "      position = blocks_snps_df_.iloc[row, 2]\n",
    "      snp_block = blocks[(blocks[\"start\"].astype(int) <= position) \n",
    "                         & (blocks[\"end\"].astype(int) >= position)][[\"chr\", \"block_id\"]]\n",
    "      snp_pos_block.append((snp_block.iloc[0][0], position, snp_block.iloc[0][1]))\n",
    "\n",
    "    snp_blocks = pd.DataFrame(snp_pos_block)\n",
    "    snp_blocks.columns = [\"chr\", \"pos\", \"block_id\"]\n",
    "\n",
    "    snp_blocks = snp_blocks.merge(positions, on=[\"chr\", \"pos\"]).drop_duplicates()\n",
    "      \n",
    "    return snp_blocks\n",
    "\n",
    "get_block_snps(callset, blocks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndentationError",
     "evalue": "expected an indented block after 'for' statement on line 4 (2163215902.py, line 7)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[130], line 7\u001b[0;36m\u001b[0m\n\u001b[0;31m    blocks_snps_df = pd.concat([positions[positions[\"pos\"].between(blocks.iloc[row, 1],\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mIndentationError\u001b[0m\u001b[0;31m:\u001b[0m expected an indented block after 'for' statement on line 4\n"
     ]
    }
   ],
   "source": [
    "positions = callset.callset_positions_df\n",
    "positions[\"gt_idx\"] = positions.index\n",
    "\n",
    "for row in range(len(blocks)):\n",
    "    \n",
    "\n",
    "blocks_snps_df = pd.concat([positions[positions[\"pos\"].between(blocks.iloc[row, 1],\n",
    "                                                                blocks.iloc[row, 2])]\n",
    "                                                                  for row in range(len(blocks))])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>block_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>73671</td>\n",
       "      <td>73772</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:73671-73772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>143980</td>\n",
       "      <td>144081</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:143980-144081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>237688</td>\n",
       "      <td>237789</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:237688-237789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>280200</td>\n",
       "      <td>280301</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:280200-280301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>352222</td>\n",
       "      <td>352323</td>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1:352222-352323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483342</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_2</td>\n",
       "      <td>1972</td>\n",
       "      <td>2073</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_2:1972-2073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483447</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_5</td>\n",
       "      <td>12075</td>\n",
       "      <td>12176</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_5:12075-12176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483477</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_5</td>\n",
       "      <td>45474</td>\n",
       "      <td>45575</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_5:45474-45575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483621</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_8</td>\n",
       "      <td>40533</td>\n",
       "      <td>40634</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_8:40533-40634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483654</th>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_9</td>\n",
       "      <td>2296</td>\n",
       "      <td>2397</td>\n",
       "      <td>brenthis_ino.SP_BI_364.scaffold_9:2296-2397</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4555 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        chr   start     end  \\\n",
       "35      brenthis_ino.SP_BI_364.chromosome_1   73671   73772   \n",
       "50      brenthis_ino.SP_BI_364.chromosome_1  143980  144081   \n",
       "127     brenthis_ino.SP_BI_364.chromosome_1  237688  237789   \n",
       "187     brenthis_ino.SP_BI_364.chromosome_1  280200  280301   \n",
       "359     brenthis_ino.SP_BI_364.chromosome_1  352222  352323   \n",
       "...                                     ...     ...     ...   \n",
       "483342    brenthis_ino.SP_BI_364.scaffold_2    1972    2073   \n",
       "483447    brenthis_ino.SP_BI_364.scaffold_5   12075   12176   \n",
       "483477    brenthis_ino.SP_BI_364.scaffold_5   45474   45575   \n",
       "483621    brenthis_ino.SP_BI_364.scaffold_8   40533   40634   \n",
       "483654    brenthis_ino.SP_BI_364.scaffold_9    2296    2397   \n",
       "\n",
       "                                                 block_id  \n",
       "35        brenthis_ino.SP_BI_364.chromosome_1:73671-73772  \n",
       "50      brenthis_ino.SP_BI_364.chromosome_1:143980-144081  \n",
       "127     brenthis_ino.SP_BI_364.chromosome_1:237688-237789  \n",
       "187     brenthis_ino.SP_BI_364.chromosome_1:280200-280301  \n",
       "359     brenthis_ino.SP_BI_364.chromosome_1:352222-352323  \n",
       "...                                                   ...  \n",
       "483342        brenthis_ino.SP_BI_364.scaffold_2:1972-2073  \n",
       "483447      brenthis_ino.SP_BI_364.scaffold_5:12075-12176  \n",
       "483477      brenthis_ino.SP_BI_364.scaffold_5:45474-45575  \n",
       "483621      brenthis_ino.SP_BI_364.scaffold_8:40533-40634  \n",
       "483654        brenthis_ino.SP_BI_364.scaffold_9:2296-2397  \n",
       "\n",
       "[4555 rows x 4 columns]"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blocks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>gt_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>7731</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>7817</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>8099</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>8157</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>8223</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33333</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>499408</td>\n",
       "      <td>33333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33334</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>499460</td>\n",
       "      <td>33334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33335</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>499498</td>\n",
       "      <td>33335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33336</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>499509</td>\n",
       "      <td>33336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33337</th>\n",
       "      <td>brenthis_ino.SP_BI_364.chromosome_1</td>\n",
       "      <td>499829</td>\n",
       "      <td>33337</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33338 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       chr     pos  gt_idx\n",
       "0      brenthis_ino.SP_BI_364.chromosome_1    7731       0\n",
       "1      brenthis_ino.SP_BI_364.chromosome_1    7817       1\n",
       "2      brenthis_ino.SP_BI_364.chromosome_1    8099       2\n",
       "3      brenthis_ino.SP_BI_364.chromosome_1    8157       3\n",
       "4      brenthis_ino.SP_BI_364.chromosome_1    8223       4\n",
       "...                                    ...     ...     ...\n",
       "33333  brenthis_ino.SP_BI_364.chromosome_1  499408   33333\n",
       "33334  brenthis_ino.SP_BI_364.chromosome_1  499460   33334\n",
       "33335  brenthis_ino.SP_BI_364.chromosome_1  499498   33335\n",
       "33336  brenthis_ino.SP_BI_364.chromosome_1  499509   33336\n",
       "33337  brenthis_ino.SP_BI_364.chromosome_1  499829   33337\n",
       "\n",
       "[33338 rows x 3 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "callset.callset_positions_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = callset.callset_positions_df\n",
    "\n",
    "df.merge(blocks, how=\"left\", on=[\"chr\", ])\n",
    "\n",
    "out = blocks.merge(df, how='left', on=\"chr\") \\\n",
    "         .query('pos.between(`start`, `end`)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "def block_snps(callset, blocks):\n",
    "    \"\"\"Merge block and callset information to get dataframe of SNPs and which block they belong to\"\"\"\n",
    "    return blocks.merge(callset.callset_positions_df, how='left', on=\"chr\") \\\n",
    "         .query('pos.between(`start`, `end`)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = block_snps(callset, blocks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert (df.columns == [\n",
    "            \"chr\", \"start\", \"end\", \"block_id\", \"pos\", \"gt_idx\"]).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chr', 'start', 'end', 'block_id', 'pos', 'gt_idx'], dtype='object')"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "block_snps_df = get_block_snps(callset, blocks)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (3) Use GT information to make a DF of calls per block"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>hap</th>\n",
       "      <th>1595</th>\n",
       "      <th>1596</th>\n",
       "      <th>1597</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SE_BI_1495</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SE_BI_1495</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FR_BI_1497</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FR_BI_1497</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FR_BD_1329</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FR_BD_1329</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>RS_BI_1496</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>RS_BI_1496</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RO_BD_956</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>RO_BD_956</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ES_BI_375</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ES_BI_375</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>UA_BI_1494</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>UA_BI_1494</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ES_BD_1141</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ES_BD_1141</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ES_BI_364</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ES_BI_364</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ES_BD_1489</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ES_BD_1489</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ES_BD_1490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ES_BD_1490</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>GR_BD_1491</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>GR_BD_1491</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>IT_BD_1493</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>IT_BD_1493</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample  hap  1595  1596  1597\n",
       "0   SE_BI_1495    0     0     0     0\n",
       "1   SE_BI_1495    1     0     0     0\n",
       "2   FR_BI_1497    0     0     0     0\n",
       "3   FR_BI_1497    1     0     0     1\n",
       "4   FR_BD_1329    0     0     1     1\n",
       "5   FR_BD_1329    1     0     1     1\n",
       "6   RS_BI_1496    0     0     0     0\n",
       "7   RS_BI_1496    1     1     0     1\n",
       "8    RO_BD_956    0     0     1     1\n",
       "9    RO_BD_956    1     0     1     1\n",
       "10   ES_BI_375    0     0     0     0\n",
       "11   ES_BI_375    1     0     0     0\n",
       "12  UA_BI_1494    0     0     0     0\n",
       "13  UA_BI_1494    1     0     0     1\n",
       "14  ES_BD_1141    0     0     1     1\n",
       "15  ES_BD_1141    1     0     1     1\n",
       "16   ES_BI_364    0     0     0     0\n",
       "17   ES_BI_364    1     0     0     0\n",
       "18  ES_BD_1489    0     0     1     1\n",
       "19  ES_BD_1489    1     0     1     1\n",
       "20  ES_BD_1490    0     0     1     1\n",
       "21  ES_BD_1490    1     0     1     1\n",
       "22  GR_BD_1491    0     0     1     1\n",
       "23  GR_BD_1491    1     0     1     1\n",
       "24  IT_BD_1493    0     0     1     1\n",
       "25  IT_BD_1493    1     0     1     1"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def block_callset(callset, gt_idxs, ploidy=2):\n",
    "    \"\"\"Get the callset for a block defined by GT indices, divided by haplotype\"\"\"\n",
    "    block_calls = pd.DataFrame(list(itertools.product(callset.samples, list(range(ploidy)))),\n",
    "                        columns=[\"sample\", \"hap\"])\n",
    "    \n",
    "    for idx in gt_idxs:\n",
    "        block_calls[idx] = callset.gt[idx].flatten()\n",
    "\n",
    "    return block_calls.dropna()\n",
    "\n",
    "block_callset_df(callset, list(range(1595, 1598)), 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "def n_segr_sites(block_callset):\n",
    "    \"\"\"Count segregating sites between samples of length 2 from a block callset\"\"\"\n",
    "    return sum(block_callset.iloc[0, 2:] != block_callset.iloc[1, 2:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "def segregating_sites_spectrum(callset, block_snps, samples_map_df, blocklen, ploidy=2, sampling_probabilities=[0.25, 0.25, 0.5]):\n",
    "\n",
    "    rng = np.random.default_rng()\n",
    "\n",
    "    populations = samples_map_df[\"population\"].unique()\n",
    "    pop1 = list(samples_map_df[\"sample\"]\n",
    "                [samples_map_df[\"population\"] == populations[0]])\n",
    "    pop2 = list(samples_map_df[\"sample\"]\n",
    "                [samples_map_df[\"population\"] == populations[1]])\n",
    "\n",
    "    sss = np.zeros(shape=(3, blocklen))\n",
    "\n",
    "    for block in block_snps[\"block_id\"].unique():\n",
    "        block_gt_idxs = list(\n",
    "            block_snps[block_snps[\"block_id\"] == block][\"gt_idx\"])\n",
    "        block_callset_df = block_callset(callset,\n",
    "                                         block_gt_idxs,\n",
    "                                         ploidy=ploidy)\n",
    "\n",
    "        sampling_state = int(np.where(rng.multinomial(\n",
    "            n=1, pvals=sampling_probabilities) == 1)[0]) + 1\n",
    "\n",
    "        if sampling_state == 1:\n",
    "            block_sss = n_segr_sites(\n",
    "                block_callset_df[block_callset_df[\"sample\"].isin(pop1)].sample(n=2))\n",
    "        elif sampling_state == 2:\n",
    "            block_sss = n_segr_sites(\n",
    "                block_callset_df[block_callset_df[\"sample\"].isin(pop2)].sample(n=2))\n",
    "        else:\n",
    "            assert sampling_state == 3\n",
    "            block_sss = n_segr_sites(pd.concat([block_callset_df[block_callset_df[\"sample\"].isin(pop1)].sample(n=1),\n",
    "                                     block_callset_df[block_callset_df[\"sample\"].isin(pop2)].sample(n=2)]))\n",
    "\n",
    "        sss[sampling_state-1, int(block_sss)\n",
    "            ] = sss[sampling_state-1, int(block_sss)] + 1\n",
    "\n",
    "    return sss\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_map_df = pd.DataFrame({\"sample\": callset.samples,\n",
    "                               \"population\": [\"BI\", \"BI\", \"BD\", \"BI\", \"BD\", \"BI\", \"BI\", \"BD\", \"BI\", \"BD\", \"BD\", \"BD\", \"BD\"]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "s3 = segregating_sites_spectrum(callset, block_snps_df, samples_map_df=samples_map_df, blocklen=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7.0"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(s3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'CallSet' object is not subscriptable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[141], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m s3 \u001b[39m=\u001b[39m segregating_sites_spectrum(\n\u001b[1;32m      2\u001b[0m     callset,\n\u001b[0;32m----> 3\u001b[0m     snp_positions\u001b[39m=\u001b[39mget_snp_positions(callset[\u001b[39m\"\u001b[39m\u001b[39mvariants/CHROM\u001b[39m\u001b[39m\"\u001b[39m][:], pos\u001b[39m=\u001b[39mcallset[\u001b[39m\"\u001b[39m\u001b[39mvariants/POS\u001b[39m\u001b[39m\"\u001b[39m][:], blocks\u001b[39m=\u001b[39mblocks),\n\u001b[1;32m      4\u001b[0m     ploidy\u001b[39m=\u001b[39m\u001b[39m2\u001b[39m,\n\u001b[1;32m      5\u001b[0m     sampling_probabilities\u001b[39m=\u001b[39m[\u001b[39m0.25\u001b[39m, \u001b[39m0.25\u001b[39m, \u001b[39m0.5\u001b[39m],\n\u001b[1;32m      6\u001b[0m     samples_df\u001b[39m=\u001b[39msamples_df,\n\u001b[1;32m      7\u001b[0m     blocklen\u001b[39m=\u001b[39m\u001b[39m100\u001b[39m\n\u001b[1;32m      8\u001b[0m )\n",
      "\u001b[0;31mTypeError\u001b[0m: 'CallSet' object is not subscriptable"
     ]
    }
   ],
   "source": [
    "s3 = segregating_sites_spectrum(\n",
    "    callset,\n",
    "    snp_positions=get_snp_positions(callset[\"variants/CHROM\"][:], pos=callset[\"variants/POS\"][:], blocks=blocks),\n",
    "    ploidy=2,\n",
    "    sampling_probabilities=[0.25, 0.25, 0.5],\n",
    "    samples_df=samples_df,\n",
    "    blocklen=100\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[444., 263., 130., 104.,  57.,  24.,  23.,  11.,  12.,   2.,   3.,\n",
       "          1.,   0.,   1.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.],\n",
       "       [429., 266., 138.,  77.,  46.,  20.,  19.,  15.,   9.,   1.,   2.,\n",
       "          0.,   0.,   0.,   0.,   1.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.],\n",
       "       [896., 451., 269., 160.,  96.,  55.,  31.,  19.,   9.,   6.,   4.,\n",
       "          1.,   2.,   1.,   0.,   1.,   1.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,\n",
       "          0.]])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def from_gff3_vcf(gff3_path,\n",
    "                  vcf_path,\n",
    "                  zarr_path,\n",
    "                  samples_map,\n",
    "                  blocklen=100,\n",
    "                    ld_dist_bp=1000,\n",
    "                    ploidy=2,\n",
    "                    sampling_probabilities=[0.25, 0.25, 0.5]):\n",
    "    \n",
    "    \"\"\"Preprocess a GFF3 annotation and VCF file to segregating sites spectrum.\"\"\"\n",
    "    \n",
    "    blocks = gff3_to_blocks(gff3_path, blocklen=blocklen, ld_dist_bp=ld_dist_bp)\n",
    "    callset = CallSet(vcf_path, zarr_path)\n",
    "    block_snps_df = block_snps(callset, blocks)\n",
    "\n",
    "    samples_map_df = pd.read_csv(samples_map)\n",
    "\n",
    "    s3 = segregating_sites_spectrum(callset,\n",
    "                                    block_snps=block_snps_df,\n",
    "                                    samples_map_df=samples_map_df,\n",
    "                                    blocklen=blocklen,\n",
    "                                    ploidy=ploidy,\n",
    "                                    sampling_probabilities=sampling_probabilities)\n",
    "    \n",
    "    return s3"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dismal",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
